src/Symfony/Component/DependencyInjection/Tests/Compiler/ValidateEnvPlaceholdersPassTest.php

```diff
@@ -252,6 +252,29 @@ public function testDiscardedEnvInConfig(): void
         $this->assertSame('1', $container->getParameter('boolish'));
     }
 
+    public function testEnvStages(): void
+    {
+        $_ENV['REAL'] = '1.5';
+        $container = new ContainerBuilder();
+        $container->setParameter('env(REAL)', 2.5);
+        $container->setParameter('env(DEFAULT)', 3.5);
+        $container->register('foo', \stdClass::class)
+            ->setPublic(true)
+            ->setProperty('real', '%env(REAL)%')
+            ->setProperty('default', '%env(DEFAULT)%');
+        $container->registerExtension(new EnvExtension());
+        $container->prependExtensionConfig('env_extension', array(
+            'float_node' => '%env(DEFAULT)%',
```

Comment:
passes validation, as it uses `float(3.5)` as is to replace the env placeholder with. BUT the failing test shows we get `string("3.5")` after compile:

```
Failed asserting that '3.5' is identical to 3.5.
```

Reply:
Why is this converted to a string? That looks weird as the node is a float node.

Reply:
Oh I see because it's read from the env var, right? So if you wanted to do it right, you would have to make use of the `float` processor? In that case I'd say having the string is expected and we should change the assertion accordingly.

Reply:
we get the string value after compile, all good there. But we get a float value during config validation (which was also string before, but broke BC => #27455

So right now we have float during config validation, but a string is actually injected. Hence my suggestion to deprecate  non-string default (which causes this discrepancy) so it's consistent again in 5.0